qiskit_experiments/data_processing/data_action.py

```diff
@@ -29,50 +31,40 @@ def __init__(self, validate: bool = True):
         """
         self._validate = validate
 
-    @abstractmethod
-    def _process(self, datum: Any, error: Optional[Any] = None) -> Tuple[Any, Any]:
-        """
-        Applies the data processing step to the datum.
+    def _process(self, gen_datum: Iterator) -> Generator:
+        """Applies the data processing step to the datum.
 
         Args:
-            datum: A single item of data which will be processed.
-            error: An optional error estimation on the datum that can be further propagated.
+            gen_datum: A generator of unprocessed data. Each entry is a tuple of data and error.
 
-        Returns:
-            processed data: The data that has been processed along with the propagated error.
+        Yields:
+            A tuple of processed data and error.
         """
+        yield from gen_datum
 
-    @abstractmethod
-    def _format_data(self, datum: Any, error: Optional[Any] = None) -> Tuple[Any, Any]:
-        """Format and validate the input.
+    def _format_data(self, gen_datum: Iterator) -> Generator:
+        """Validate and format the input.
 
-        Check that the given data and error has the correct structure. This method may
-        additionally change the data type, e.g. converting a list to a numpy array.
+        Check that the given data and error have the correct structure.
 
         Args:
-            datum: The data instance to check and format.
-            error: An optional error estimation on the datum to check and format.
-
-        Returns:
-            datum, error: The formatted datum and its optional error.
+            gen_datum: A generator of unformatted data. Each entry is a tuple of data and error.
 
-        Raises:
-            DataProcessorError: If either the data or the error do not have the proper format.
+        Yields:
+            A tuple of formatted data and error.
         """
+        yield from gen_datum
 
-    def __call__(self, data: Any, error: Optional[Any] = None) -> Tuple[Any, Any]:
+    def __call__(self, gen_datum: Iterator) -> Generator:
```

Comment:
Please add more documentation to explain how this functions. With this change the interface is becoming more abstract and harder to follow. This therefore requires more careful explanation of how the `DataAction` functions.